src/integration/java/org/springframework/data/couchbase/repository/SimpleCouchbaseRepositoryTests.java

```diff
@@ -220,6 +223,40 @@ public void shouldTakeVersionIntoAccountWhenDoingMultipleUpdates() {
     }
   }
 
+  @Test
+  public void shouldFailOnMultipleConcurrentSaves() throws Exception {
+    final String key = "shouldFailOnMultipleConcurrentSaves";
+    final AtomicLong counter = new AtomicLong();
+    final AtomicLong optimisticLockCounter = new AtomicLong();
+    ExecutorService pool = Executors.newFixedThreadPool(5);
+
+    Callable<Void> task = new Callable<Void>() {
+      @Override
+      public Void call() throws Exception {
+        long counterValue = counter.incrementAndGet();
+        String data = "value-" + counterValue;
+        VersionedData messageData = new VersionedData(key, data);
+        try {
+          versionedDataRepository.save(messageData);
+        } catch (OptimisticLockingFailureException e) {
+          optimisticLockCounter.incrementAndGet();
+        }
+        //should save operation throw OptimisticLockingFailureException on next attempts to save?
+        return null;
+      }
+    };
+    Collection<Callable<Void>> tasks = Collections.nCopies(5, task);
+
+    List<Future<Void>> futures = pool.invokeAll(tasks);
+    for (Future future : futures) {
+      future.get(5, TimeUnit.SECONDS);
+    }
+
+    versionedDataRepository.delete(key);
```

Comment:
shouldn't the delete be made at the beginning of test? maybe test both cases (one where a pre-existing version of the doc exists => update, the other where it is `deleteIfExist` => insert)?